### PR TITLE
Fix mistranslation of side-effects inside unused unary operations

### LINKED ISSUE #896: c2rust-transpile: bug about side effect of expression
c2rust v0.17.0
Ubuntu 20.04.2

Consider the following C programs:
```c
// output:
// global_v = 0
// global_v = 1

// C programs
#include <stdio.h>
int global_v = 0;
int func();
int main(){ 
    printf("global_v = %d\n", global_v); 
    // the result of func() is not used, but necessary
    -func();
    printf("global_v = %d\n", global_v); 
}

int func() { 
    // change the value of global_v
    global_v = 1; 
    return global_v;
}
```
the generated Rust program ignore to translate the expression: "-func();", which make the different output between C and Rust.
```rust
// output:
// global_v = 0
// global_v = 0

// Rust program
use ::correctness_1::*;
extern "C" {
    fn printf(_: *const libc::c_char, _: ...) -> libc::c_int;
}
#[no_mangle]
pub static mut global_v: libc::c_int = 0 as libc::c_int;
unsafe fn main_0() -> libc::c_int {
    printf(b"global_v = %d\n\0" as *const u8 as *const libc::c_char, global_v);
    // ignore the expression: "-func();"
    printf(b"global_v = %d\n\0" as *const u8 as *const libc::c_char, global_v);
    return 0;
}
#[no_mangle]
pub unsafe extern "C" fn func() -> libc::c_int {
    global_v = 1 as libc::c_int;
    return global_v;
}
pub fn main() {
    unsafe { ::std::process::exit(main_0() as i32) }
}
```
however, when the C code just call func() as the following:
```c
int main(){ 
    printf("global_v = %d\n", global_v); 
    // the result of func() is not used, but necessary
    func();
    printf("global_v = %d\n", global_v); 
}
```
the translated Rust code is correct:
```rust
unsafe fn main_0() -> libc::c_int {
    printf(b"global_v = %d\n\0" as *const u8 as *const libc::c_char, global_v);
    func();
    printf(b"global_v = %d\n\0" as *const u8 as *const libc::c_char, global_v);
    return 0;
}
```

So, the error occurs in c2rust/c2rust-transpile/src/translator/operators.rs
the following code about `Unop` doesn't consider about side effect of expression:
```rust
            // c2rust/c2rust-transpile/src/translator/operators.rs 1004-1029
            c_ast::UnOp::Plus => self.convert_expr(ctx.used(), arg), // promotion is explicit in the clang AST

            c_ast::UnOp::Negate => {
                let val = self.convert_expr(ctx.used(), arg)?;

                if resolved_ctype.kind.is_unsigned_integral_type() {
                    Ok(val.map(wrapping_neg_expr))
                } else {
                    Ok(val.map(neg_expr))
                }
            }
            c_ast::UnOp::Complement => Ok(self
                .convert_expr(ctx.used(), arg)?
                .map(|a| mk().unary_expr(UnOp::Not(Default::default()), a))),

            c_ast::UnOp::Not => {
                let val = self.convert_condition(ctx, false, arg)?;
                Ok(val.map(|x| mk().cast_expr(x, mk().path_ty(vec!["libc", "c_int"]))))
            }
            c_ast::UnOp::Extension => {
                let arg = self.convert_expr(ctx, arg)?;
                Ok(arg)
            }
```
could chage like the following, thus it can correctly consider about side effcet: 
```rust
           c_ast::UnOp::Plus => {
                let mut res = self.convert_expr(ctx.used(), arg)?;
                if ctx.is_unused() {
                    let v = res.clone().into_value();
                    res.stmts_mut().push(Stmt::Semi(*v, Default::default()));
                }
                Ok(res)
            } // promotion is explicit in the clang AST

            c_ast::UnOp::Negate => {
                let val = self.convert_expr(ctx.used(), arg)?;

                let mut res = if resolved_ctype.kind.is_unsigned_integral_type() {
                    val.map(wrapping_neg_expr)
                } else {
                    val.map(neg_expr)
                };
                if ctx.is_unused() {
                    let v = res.clone().into_value();
                    res.stmts_mut().push(Stmt::Semi(*v, Default::default()));
                }
                Ok(res)
            }
            c_ast::UnOp::Complement => {
                let mut res = self
                    .convert_expr(ctx.used(), arg)?
                    .map(|a| mk().unary_expr(UnOp::Not(Default::default()), a));
                if ctx.is_unused() {
                    let v = res.clone().into_value();
                    res.stmts_mut().push(Stmt::Semi(*v, Default::default()));
                }
                Ok(res)
            }

            c_ast::UnOp::Not => {
                let val = self.convert_condition(ctx, false, arg)?;
                let mut res = val.map(|x| mk().cast_expr(x, mk().path_ty(vec!["libc", "c_int"])));
                if ctx.is_unused() {
                    let v = res.clone().into_value();
                    res.stmts_mut().push(Stmt::Semi(*v, Default::default()));
                }
                Ok(res)
            }
            c_ast::UnOp::Extension => {
                let arg = self.convert_expr(ctx, arg)?;
                Ok(arg)
            }
```

### PR DESCRIPTION
**Pull request by fw-immunant**
Fixes #896.

### CODE REVIEW

File: tests/misc/src/test_exprs.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+//! feature_core_intrinsics,
```
Comment (kkysen) [inline tests/misc/src/test_exprs.rs:1]: Why do we need `#![feature(core_intrinsics)]` for this?

File: tests/misc/src/exprs.c [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+void exprs(const unsigned n, int * const buffer) {
+
+        unsigned long i = 0;
+		char arr[1] = {0};
+
+        side_effect(&buffer[0]);
+        -side_effect(&buffer[1]);
+        &""[side_effect(&buffer[2])];
+		++arr[side_effect(&buffer[3])];
+}
```
Comment (kkysen) [inline tests/misc/src/exprs.c:17]: nit: Indentation.

File: c2rust-transpile/src/translator/operators.rs [hunk 3]
```diff
@@ -944,6 +944,14 @@ impl<'c> Translation<'c> {
             c_ast::UnOp::Real | c_ast::UnOp::Imag | c_ast::UnOp::Coawait => {
                 panic!("Unsupported extension operator")
             }
+        }?;
+
+        // Unused unary operators (`-foo()`) may have side effects, so we need
```
Comment (kkysen) [inline c2rust-transpile/src/translator/operators.rs:949]:
```suggestion
        // Unused unary operators (e.x. `-foo()`) may have side effects, so we need
```

**Review by kkysen — APPROVED** [review]
LGTM besides the little things I commented on.  Thanks, @zone-hai, for finding the bug and the clear fix!

File: tests/misc/src/test_exprs.rs (hunk 1 above)
Comment (kkysen) [reply tests/misc/src/test_exprs.rs:1]: I think it does need `//! feature_label_break_value`, though, as `#![feature(label_break_value)]` was only recently stabilized and we're still stuck on an older `rustc`.

Comment (fw-immunant) [reply tests/misc/src/test_exprs.rs:1]: We don't--it was a holdover from copy-pasting another test case. Removed.

Comment (fw-immunant) [reply tests/misc/src/test_exprs.rs]: Added `feature_label_break_value` for the `sizeof` test. I'm not sure why that one wasn't failing in CI prior to now.

File: tests/misc/src/exprs.c (hunk 2 above)
Comment (kkysen) [reply tests/misc/src/exprs.c]: I meant the 8-space indentation.

File: tests/misc/src/test_exprs.rs (hunk 1 above)
Comment (kkysen) [reply tests/misc/src/test_exprs.rs]: Yeah, that's weird, but I guess if it's fixed now it's all good.

File: tests/misc/src/exprs.c (hunk 2 above)
Comment (fw-immunant) [reply tests/misc/src/exprs.c]: I was following the pattern of `sizeofs.c`. We should normalize all these, but I didn't do a survey to see what was most common.

Comment (kkysen) [reply tests/misc/src/exprs.c]: Ahh, okay.

**Pull request merged**